### PR TITLE
KAFKA-6748: double check before scheduling a new task after the punctuate call

### PR DESCRIPTION
**Pull request by fredfp**
The first commit is a new test, illustrating the problem. The second commit fixes the problem.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by fredfp** [comment]
I can't add any watchers to the JIRA issue, so here we go.

@guozhangwang this is something I already noticed and we discussed [earlier in the ML](http://mail-archives.apache.org/mod_mbox/kafka-users/201712.mbox/%3CCAGKO5_4UcygeKMVYtfx3rj6sOLUm_JUqsZn2b4bZMsH3izKpRA%40mail.gmail.com%3E). I got bitten by this again and investigated further.

@mjsax Could you review please?

**Comment by mjsax** [comment]
\cc @bbejeck @vvcephei

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/PunctuationQueueTest.java [hunk 1]
```diff
@@ -126,6 +127,45 @@ public void punctuate(ProcessorNode node, long time, PunctuationType type, Punct
...
+        queue.mayPunctuate(now, PunctuationType.STREAM_TIME, processorNodePunctuator);
+        assertEquals(0, processor.punctuatedAt.size());
+
+        queue.mayPunctuate(now + 100L, PunctuationType.STREAM_TIME, processorNodePunctuator);
+        assertEquals(1, processor.punctuatedAt.size());
+
+        queue.mayPunctuate(now + 200L, PunctuationType.STREAM_TIME, processorNodePunctuator);
+        assertEquals(1, processor.punctuatedAt.size());
+
+        queue.mayPunctuate(now + 1001L, PunctuationType.STREAM_TIME, processorNodePunctuator);
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/PunctuationQueueTest.java]: Why we want to test `mayPunctuate(now + 1001L ..)` again here?

**Review by guozhangwang — COMMENTED** [review]
LGTM! Thanks for the catch.

Just a minor question about the test function.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/PunctuationQueueTest.java (hunk 1 above)
Comment (fredfp) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/PunctuationQueueTest.java]: You are right, I just wanted to be 100% sure when writing the test, forgot to remove...

**Review by vvcephei — APPROVED** [review]
Thanks for the PR!

**Comment by guozhangwang** [comment]
Also cherry-picking to 1.1.

**Comment by fredfp** [comment]
@guozhangwang Thanks! What about 1.0?

**Comment by guozhangwang** [comment]
I have cherry-picked to 1.0 as well, but we do not have another bugfix release plan for 1.x atm, just FYI. If there is a 1.0.2 bug fix release coming up later, it will include this fix.

**Pull request merged**